kernel/comps/block/src/partition.rs

```diff
@@ -0,0 +1,298 @@
+    let mut partitions = Vec::new();
+    let mut extended_partition = None;
+    for entry in mbr.entries {
+        if entry.is_extended() {
+            extended_partition = Some(entry.start_sector);
+        }
+
+        if entry.is_valid() {
+            partitions.push(Some(PartitionInfo::Mbr(entry)));
+        } else {
+            partitions.push(None);
+        }
+    }
+
+    if let Some(start_sector) = extended_partition {
+        parse_ebr(device, &mut partitions, start_sector, 0);
+    }
+
+    partitions
+}
+
+fn parse_ebr(
+    device: &Arc<dyn BlockDevice>,
+    partitions: &mut Vec<Option<PartitionInfo>>,
+    start_sector: u32,
+    offset: u32,
+) {
+    let ebr_sector = start_sector + offset;
+    let mut ebr = device
+        .read_val::<MbrHeader>(ebr_sector as usize * SECTOR_SIZE)
+        .unwrap();
+    if ebr.entries[0].is_valid() {
+        ebr.entries[0].start_sector += ebr_sector;
+        partitions.push(Some(PartitionInfo::Mbr(ebr.entries[0])));
+    }
+
+    if ebr.entries[1].is_extended() {
+        parse_ebr(
+            device,
+            partitions,
+            start_sector,
+            ebr.entries[1].start_sector,
+        );
+    }
+}
+
+fn parse_gpt(device: &Arc<dyn BlockDevice>) -> Vec<Option<PartitionInfo>> {
+    let mut partitions = Vec::new();
+
+    // The primary GPT Header must be located in LBA 1.
+    let gpt = device.read_val::<GptHeader>(SECTOR_SIZE).unwrap();
+
+    if !gpt.check_signature() {
+        return partitions;
+    }
+
+    // TODO: Check the CRC32 of the header and the partition entries, check the backup GPT header.
+
+    let entry_size = gpt.size_of_partition_entry as usize;
+    let entries_per_sector = SECTOR_SIZE / entry_size;
+    let total_sectors = gpt.nr_partition_entries as usize / entries_per_sector;
+    for i in 0..total_sectors {
+        let mut buf = [0u8; SECTOR_SIZE];
+        let offset = (gpt.partition_entry_lba as usize + i) * SECTOR_SIZE;
+        device.read_bytes(offset, buf.as_mut_slice()).unwrap();
+
+        for j in 0..entries_per_sector {
+            let entry_offset = j * gpt.size_of_partition_entry as usize;
+            let entry = GptEntry::from_bytes(&buf[entry_offset..entry_offset + entry_size]);
+            if entry.is_valid() {
+                partitions.push(Some(PartitionInfo::Gpt(entry)));
+            } else {
+                partitions.push(None);
+            }
+        }
+    }
+
+    partitions
+}
+
+#[derive(Debug)]
+pub struct PartitionNode {
+    id: DeviceId,
+    name: String,
+    device: Arc<dyn BlockDevice>,
+    info: PartitionInfo,
+}
+
+impl BlockDevice for PartitionNode {
+    fn enqueue(&self, bio: SubmittedBio) -> Result<(), BioEnqueueError> {
+        // FIXME: This function should not be blocked by the new bio,
+        // and cloning the entire block I/O request is too inefficient.
+        let start_sid = bio.sid_range().start + self.info.start_sector();
```

Comment:
This is an important limitation and it deserves to be fixed---in this PR. What solutions have you explored? And why all of them are not working?